File: core/play/src/main/scala/play/api/mvc/WebSocket.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.mvc

import scala.concurrent.Future
import scala.util.control.NonFatal

import org.apache.pekko.actor.ActorRef
import org.apache.pekko.actor.Props
import org.apache.pekko.stream.scaladsl.Flow
import org.apache.pekko.util.ByteString
import play.api.http.websocket._
import play.api.http.HeaderNames
import play.api.libs.json._
import play.api.libs.streams.PekkoStreams
import play.core.Execution.Implicits.trampoline

/**
 * A WebSocket handler.
 */
trait WebSocket extends Handler {

  /**
   * Execute the WebSocket.
   *
   * The return value is either a result to reject the WebSocket with (or otherwise respond in a different way), or
   * a flow to handle the WebSocket messages.
   */
  def apply(request: RequestHeader): Future[Either[Result, Flow[Message, Message, ?]]]

  /**
   * Execute the WebSocket, including WebSocket handshake metadata.
   *
   * The return value is either a result to reject the WebSocket with (or otherwise respond in a different way), or
   * an accepted WebSocket containing a flow to handle the WebSocket messages and optional handshake metadata.
   */
  def applyWithOptions(request: RequestHeader): Future[Either[Result, WebSocket.Accepted[Message, Message]]] = {
    apply(request).map(_.map(flow => WebSocket.Accepted(flow, WebSocket.firstRequestedSubprotocol(request))))
  }
}

/**
 * Helper utilities to generate WebSocket results.
 */
object WebSocket {
  private val ReservedHandshakeResponseHeaders = Set(
    "connection",
    "sec-websocket-accept",
    "sec-websocket-extensions",
    "sec-websocket-protocol",
    "upgrade"
  )

  private[play] def firstRequestedSubprotocol(request: RequestHeader): Option[String] = {
    request.headers
      .get("Sec-WebSocket-Protocol")
      .toSeq
      .flatMap(_.split(",").iterator.map(_.trim).filter(_.nonEmpty))
      .headOption
  }

  private[play] def allowedHandshakeResponseHeaders(headers: Headers): Seq[(String, String)] = {
    headers.headers.filterNot {
      case (name, _) =>
        ReservedHandshakeResponseHeaders(name.toLowerCase(java.util.Locale.ROOT))
    }
  }

  def apply(f: RequestHeader => Future[Either[Result, Flow[Message, Message, ?]]]): WebSocket = {
    (request: RequestHeader) => f(request)
  }

  /**
   * An accepted WebSocket, including the flow that handles WebSocket messages and optional handshake metadata.
   *
   * @param flow the flow that handles WebSocket messages
   * @param subprotocol the WebSocket subprotocol selected by the application, if any
   * @param headers additional HTTP headers to send with the WebSocket upgrade response
   * @param newCookies additional cookies to send with the WebSocket upgrade response
   */
  case class Accepted[In, Out](
      flow: Flow[In, Out, ?],
      subprotocol: Option[String] = None,
      headers: Headers = Headers(),
      newCookies: Seq[Cookie] = Seq.empty
  ) {

    /**
     * Adds headers to this WebSocket upgrade response.
     */
    def withHeaders(headers: (String, String)*): Accepted[In, Out] = {
      copy(headers = this.headers.add(headers*))
    }

    /**
     * Discards a header from this WebSocket upgrade response.
     */
    def discardingHeader(name: String): Accepted[In, Out] = {
      copy(headers = headers.remove(name))
    }

    /**
     * Adds cookies to this WebSocket upgrade response. If the response already contains cookies then cookies with the
     * same name in the new list will override existing ones.
     */
    def withCookies(cookies: Cookie*): Accepted[In, Out] = {
      val filteredCookies = newCookies.filter(cookie => !cookies.exists(_.name == cookie.name))
      if (cookies.isEmpty) this else copy(newCookies = filteredCookies ++ cookies)
    }

    /**
     * Discards cookies along this WebSocket upgrade response.
     */
    def discardingCookies(cookies: DiscardingCookie*): Accepted[In, Out] = {
      withCookies(cookies.map(_.toCookie)*)
    }

    private[play] def bakeCookies(
        cookieHeaderEncoding: CookieHeaderEncoding = new DefaultCookieHeaderEncoding()
    ): Accepted[In, Out] = {
      val allCookies = {
        val setCookieCookies =
          cookieHeaderEncoding.decodeSetCookieHeader(headers.get(HeaderNames.SET_COOKIE).getOrElse(""))
        setCookieCookies ++ newCookies
      }

      if (allCookies.isEmpty) {
        this
      } else {
        withHeaders(HeaderNames.SET_COOKIE -> cookieHeaderEncoding.encodeSetCookieHeader(allCookies))
      }
    }
  }

  /**
   * Transforms WebSocket message flows into message flows of another type.
   *
   * The transformation may be more than just converting from one message to another, it may also produce messages, such
   * as close messages with an appropriate error code if the message can't be consumed.
   */
  trait MessageFlowTransformer[+In, -Out] { self =>

    /**
     * Transform the flow of In/Out messages into a flow of WebSocket messages.
     */
    def transform(flow: Flow[In, Out, ?]): Flow[Message, Message, ?]

    /**
     * Contramap the out type of this transformer.
     */
    def contramap[NewOut](f: NewOut => Out): MessageFlowTransformer[In, NewOut] = { (flow: Flow[In, NewOut, ?]) =>
      self.transform(
        flow.map(f)
      )
    }

    /**
     * Map the in type of this transformer.
     */
    def map[NewIn](f: In => NewIn): MessageFlowTransformer[NewIn, Out] = { (flow: Flow[NewIn, Out, ?]) =>
      self.transform(
        Flow[In].map(f).via(flow)
      )
    }

    /**
     * Map the in type and contramap the out type of this transformer.
     */
    def map[NewIn, NewOut](f: In => NewIn, g: NewOut => Out): MessageFlowTransformer[NewIn, NewOut] = {
      (flow: Flow[NewIn, NewOut, ?]) =>
        {
          self.transform(
            Flow[In].map(f).via(flow).map(g)
          )
        }
    }
  }

  object MessageFlowTransformer {
    implicit val identityMessageFlowTransformer: MessageFlowTransformer[Message, Message] = {
      (flow: Flow[Message, Message, ?]) => flow
    }

    /**
     * Converts text messages to/from Strings.
     */
    implicit val stringMessageFlowTransformer: MessageFlowTransformer[String, String] = {
      (flow: Flow[String, String, ?]) =>
        {
          PekkoStreams.bypassWith[Message, String, Message](Flow[Message].collect {
            case TextMessage(text) => Left(text)
            case BinaryMessage(_)  =>
              Right(CloseMessage(Some(CloseCodes.Unacceptable), "This WebSocket only supports text frames"))
          })(flow.map(TextMessage.apply))
        }
    }

    /**
     * Converts binary messages to/from ByteStrings.
     */
    implicit val byteStringMessageFlowTransformer: MessageFlowTransformer[ByteString, ByteString] = {
      (flow: Flow[ByteString, ByteString, ?]) =>
        {
          PekkoStreams.bypassWith[Message, ByteString, Message](Flow[Message].collect {
            case BinaryMessage(data) => Left(data)
            case TextMessage(_)      =>
              Right(CloseMessage(Some(CloseCodes.Unacceptable), "This WebSocket only supports binary frames"))
          })(flow.map(BinaryMessage.apply))
        }
    }

    /**
     * Converts binary messages to/from byte arrays.
     */
    implicit val byteArrayMessageFlowTransformer: MessageFlowTransformer[Array[Byte], Array[Byte]] = {
      byteStringMessageFlowTransformer.map(_.toArray, ByteString.apply)
    }

    /**
     * Converts messages to/from JsValue
     */
    implicit val jsonMessageFlowTransformer: MessageFlowTransformer[JsValue, JsValue] = {
      def closeOnException[T](block: => T) =
        try {
          Left(block)
        } catch {
          case NonFatal(e) => Right(CloseMessage(Some(CloseCodes.Unacceptable), "Unable to parse json message"))
        }

      (flow: Flow[JsValue, JsValue, ?]) => {
        PekkoStreams.bypassWith[Message, JsValue, Message](Flow[Message].collect {
          case BinaryMessage(data) => closeOnException(Json.parse(data.asInputStream))
          case TextMessage(text)   => closeOnException(Json.parse(text))
        })(flow.map { json => TextMessage(Json.stringify(json)) })
      }
    }

    /**
     * Converts messages to/from a JSON high level object.
     *
     * If the input messages fail to be parsed, the WebSocket will be closed with an 1003 close code and the parse error
     * serialised to JSON.
     */
    def jsonMessageFlowTransformer[In: Reads, Out: Writes]: MessageFlowTransformer[In, Out] = {
      jsonMessageFlowTransformer.map(
        json =>
          Json
            .fromJson[In](json)
            .fold(
              { errors =>
                throw WebSocketCloseException(
                  CloseMessage(Some(CloseCodes.Unacceptable), Json.stringify(JsError.toJson(errors)))
                )
              },
              identity
            ),
        out => Json.toJson(out)
      )
    }
  }

  /**
   * Accepts a WebSocket using the given flow.
   */
  def accept[In, Out](
      f: RequestHeader => Flow[In, Out, ?]
  )(implicit transformer: MessageFlowTransformer[In, Out]): WebSocket = {
    acceptOrResult(f.andThen(flow => Future.successful(Right(flow))))
  }

  /**
   * Accepts a WebSocket using the given flow and handshake metadata.
   */
  def acceptWithOptions[In, Out](
      f: RequestHeader => Accepted[In, Out]
  )(implicit transformer: MessageFlowTransformer[In, Out]): WebSocket = {
    acceptOrResultWithOptions(f.andThen(accepted => Future.successful(Right(accepted))))
  }

  /**
   * Creates an action that will either accept the websocket, using the given flow to handle the in and out stream, or
   * return a result to reject the Websocket.
   */
  def acceptOrResult[In, Out](
      f: RequestHeader => Future[Either[Result, Flow[In, Out, ?]]]
  )(implicit transformer: MessageFlowTransformer[In, Out]): WebSocket = {
    WebSocket { request => f(request).map(_.map(transformer.transform)) }
  }

  /**
   * Creates an action that will either accept the websocket, using the given flow and handshake metadata, or return a
   * result to reject the Websocket.
   */
  def acceptOrResultWithOptions[In, Out](
      f: RequestHeader => Future[Either[Result, Accepted[In, Out]]]
  )(implicit transformer: MessageFlowTransformer[In, Out]): WebSocket = {
    new WebSocket {
      override def apply(request: RequestHeader): Future[Either[Result, Flow[Message, Message, ?]]] = {
        applyWithOptions(request).map(_.map(_.flow))
      }

      override def applyWithOptions(request: RequestHeader): Future[Either[Result, Accepted[Message, Message]]] = {
        f(request).map(_.map { accepted =>
          Accepted(transformer.transform(accepted.flow), accepted.subprotocol, accepted.headers, accepted.newCookies)
        })
      }
    }
  }

  /**
   * A function that, given an actor to send upstream messages to, returns actor props to create an actor to handle
   * the WebSocket
   */
  type HandlerProps = ActorRef => Props
}

File: core/play/src/main/java/play/mvc/WebSocket.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.mvc;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.pekko.stream.javadsl.Flow;
import org.apache.pekko.util.ByteString;
import play.api.http.websocket.CloseCodes;
import play.http.websocket.Message;
import play.libs.F;
import play.libs.Scala;
import play.libs.streams.PekkoStreams;
import scala.PartialFunction;

/** A WebSocket handler. */
public abstract class WebSocket {

  /**
   * Invoke the WebSocket.
   *
   * @param request The request for the WebSocket.
   * @return A future of either a result to reject the WebSocket connection with, or a Flow to
   *     handle the WebSocket.
   */
  public abstract CompletionStage<F.Either<Result, Flow<Message, Message, ?>>> apply(
      Http.RequestHeader request);

  /**
   * Invoke the WebSocket, including WebSocket handshake metadata.
   *
   * @param request The request for the WebSocket.
   * @return A future of either a result to reject the WebSocket connection with, or an accepted
   *     WebSocket with the flow and handshake metadata.
   */
  public CompletionStage<F.Either<Result, Accepted<Message, Message>>> applyWithOptions(
      Http.RequestHeader request) {
    return apply(request)
        .thenApply(
            resultOrFlow -> {
              if (resultOrFlow.left.isPresent()) {
                return F.Either.Left(resultOrFlow.left.get());
              } else {
                return F.Either.Right(
                    new Accepted<>(resultOrFlow.right.get(), firstRequestedSubprotocol(request)));
              }
            });
  }

  /** Acceptor for WebSockets to directly handle Play's Message objects. */
  public static final MappedWebSocketAcceptor<Message, Message> Message =
      new WebSocket.MappedWebSocketAcceptor<>(
          Scala.partialFunction(message -> F.Either.Left(message)), Function.identity());

  /** Acceptor for text WebSockets. */
  public static final MappedWebSocketAcceptor<String, String> Text =
      new MappedWebSocketAcceptor<>(
          Scala.partialFunction(
              message -> {
                if (message instanceof Message.Text) {
                  return F.Either.Left(((Message.Text) message).data());
                } else if (message instanceof Message.Binary) {
                  return F.Either.Right(
                      new Message.Close(
                          CloseCodes.Unacceptable(), "This websocket only accepts text frames"));
                } else {
                  throw Scala.noMatch();
                }
              }),
          Message.Text::new);

  /** Acceptor for binary WebSockets. */
  public static final MappedWebSocketAcceptor<ByteString, ByteString> Binary =
      new MappedWebSocketAcceptor<>(
          Scala.partialFunction(
              message -> {
                if (message instanceof Message.Binary) {
                  return F.Either.Left(((Message.Binary) message).data());
                } else if (message instanceof Message.Text) {
                  return F.Either.Right(
                      new Message.Close(
                          CloseCodes.Unacceptable(), "This websocket only accepts binary frames"));
                } else {
                  throw Scala.noMatch();
                }
              }),
          Message.Binary::new);

  /** Acceptor for JSON WebSockets. */
  public static final MappedWebSocketAcceptor<JsonNode, JsonNode> Json =
      new MappedWebSocketAcceptor<>(
          Scala.partialFunction(
              message -> {
                try {
                  if (message instanceof Message.Binary) {
                    return F.Either.Left(
                        play.libs.Json.parse(((Message.Binary) message).data().asInputStream()));
                  } else if (message instanceof Message.Text) {
                    return F.Either.Left(play.libs.Json.parse(((Message.Text) message).data()));
                  }
                } catch (RuntimeException e) {
                  return F.Either.Right(
                      new Message.Close(CloseCodes.Unacceptable(), "Unable to parse JSON message"));
                }
                throw Scala.noMatch();
              }),
          json -> new Message.Text(play.libs.Json.stringify(json)));

  /**
   * Acceptor for JSON WebSockets.
   *
   * @param in The class of the incoming messages, used to decode them from the JSON.
   * @param <In> The websocket's input type (what it receives from clients)
   * @param <Out> The websocket's output type (what it writes to clients)
   * @return The WebSocket acceptor.
   */
  public static <In, Out> MappedWebSocketAcceptor<In, Out> json(Class<In> in) {
    return new MappedWebSocketAcceptor<>(
        Scala.partialFunction(
            message -> {
              try {
                if (message instanceof Message.Binary) {
                  return F.Either.Left(
                      play.libs.Json.mapper()
                          .readValue(((Message.Binary) message).data().asInputStream(), in));
                } else if (message instanceof Message.Text) {
                  return F.Either.Left(
                      play.libs.Json.mapper().readValue(((Message.Text) message).data(), in));
                }
              } catch (Exception e) {
                return F.Either.Right(new Message.Close(CloseCodes.Unacceptable(), e.getMessage()));
              }
              throw Scala.noMatch();
            }),
        outMessage -> {
          try {
            return new Message.Text(play.libs.Json.mapper().writeValueAsString(outMessage));
          } catch (Exception e) {
            throw new RuntimeException(e);
          }
        });
  }

  /**
   * An accepted WebSocket, including the flow that handles WebSocket messages and optional
   * handshake metadata.
   *
   * @param <In> the type the websocket reads from clients
   * @param <Out> the type the websocket outputs back to remote clients
   */
  public static class Accepted<In, Out> {
    private final Flow<In, Out, ?> flow;
    private final Optional<String> subprotocol;
    private final List<Map.Entry<String, String>> headers;
    private final List<Http.Cookie> cookies;
    private final Http.Session session;

    public Accepted(Flow<In, Out, ?> flow, Optional<String> subprotocol) {
      this(flow, subprotocol, Collections.emptyList(), Collections.emptyList());
    }

    public Accepted(
        Flow<In, Out, ?> flow,
        Optional<String> subprotocol,
        List<Map.Entry<String, String>> headers,
        List<Http.Cookie> cookies) {
      this(flow, subprotocol, headers, cookies, null);
    }

    public Accepted(
        Flow<In, Out, ?> flow,
        Optional<String> subprotocol,
        List<Map.Entry<String, String>> headers,
        List<Http.Cookie> cookies,
        Http.Session session) {
      this.flow = flow;
      this.subprotocol = subprotocol;
      this.headers = Collections.unmodifiableList(new ArrayList<>(headers));
      this.cookies = Collections.unmodifiableList(new ArrayList<>(cookies));
      this.session = session;
    }

    public Accepted(Flow<In, Out, ?> flow, String subprotocol) {
      this(flow, Optional.of(subprotocol));
    }

    public Accepted(Flow<In, Out, ?> flow) {
      this(flow, Optional.empty());
    }

    public Flow<In, Out, ?> flow() {
      return flow;
    }

    public Optional<String> subprotocol() {
      return subprotocol;
    }

    public List<Map.Entry<String, String>> headers() {
      return headers;
    }

    public List<Http.Cookie> cookies() {
      return cookies;
    }

    public Optional<Http.Session> session() {
      return Optional.ofNullable(session);
    }

    /**
     * Return a copy of this accepted WebSocket with the given handshake response header.
     *
     * @param name the header name
     * @param value the header value
     * @return the transformed copy
     */
    public Accepted<In, Out> withHeader(String name, String value) {
      List<Map.Entry<String, String>> newHeaders = new ArrayList<>(headers);
      newHeaders.add(new AbstractMap.SimpleImmutableEntry<>(name, value));
      return new Accepted<>(flow, subprotocol, newHeaders, cookies, session);
    }

    /**
     * Return a copy of this accepted WebSocket with the given handshake response headers.
     *
     * <p>The headers are processed in pairs, so nameValues(0) is the first header's name, and
     * nameValues(1) is the first header's value, nameValues(2) is second header's name, and so on.
     *
     * @param nameValues the array of names and values.
     * @return the transformed copy
     */
    public Accepted<In, Out> withHeaders(String... nameValues) {
      if (nameValues.length % 2 != 0) {
        throw new IllegalArgumentException(
            "Headers must be supplied as alternating name and value strings");
      }

      List<Map.Entry<String, String>> newHeaders = new ArrayList<>(headers);
      for (int i = 0; i < nameValues.length; i += 2) {
        newHeaders.add(new AbstractMap.SimpleImmutableEntry<>(nameValues[i], nameValues[i + 1]));
      }
      return new Accepted<>(flow, subprotocol, newHeaders, cookies, session);
    }

    /**
     * Discard a handshake response header.
     *
     * @param name the header name
     * @return the transformed copy
     */
    public Accepted<In, Out> withoutHeader(String name) {
      String lowerName = name.toLowerCase(Locale.ROOT);
      List<Map.Entry<String, String>> newHeaders =
          headers.stream()
              .filter(header -> !header.getKey().toLowerCase(Locale.ROOT).equals(lowerName))
              .collect(Collectors.toList());
      return new Accepted<>(flow, subprotocol, newHeaders, cookies, session);
    }

    /**
     * Returns a copy of this accepted WebSocket with the given cookies.
     *
     * @param newCookies the cookies to add to the handshake response.
     * @return the transformed copy.
     */
    public Accepted<In, Out> withCookies(Http.Cookie... newCookies) {
      List<Http.Cookie> finalCookies =
          Stream.concat(
                  cookies.stream()
                      .filter(
                          cookie -> {
                            for (Http.Cookie newCookie : newCookies) {
                              if (cookie.name().equals(newCookie.name())) return false;
                            }
                            return true;
                          }),
                  Stream.of(newCookies))
              .collect(Collectors.toList());
      return new Accepted<>(flow, subprotocol, headers, finalCookies, session);
    }

    /**
     * Discard a cookie on the default path ("/") with no domain and that's not secure.
     *
     * @param name The name of the cookie to discard, must not be null
     */
    public Accepted<In, Out> discardingCookie(String name) {
      return discardingCookie(name, "/");
    }

    /**
     * Discard a cookie on the given path with no domain and not that's secure.
     *
     * @param name The name of the cookie to discard, must not be null
     * @param path The path of the cookie to discard, may be null
     */
    public Accepted<In, Out> discardingCookie(String name, String path) {
      return discardingCookie(name, path, null);
    }

    /**
     * Discard a cookie on the given path and domain that's not secure.
     *
     * @param name The name of the cookie to discard, must not be null
     * @param path The path of the cookie te discard, may be null
     * @param domain The domain of the cookie to discard, may be null
     */
    public Accepted<In, Out> discardingCookie(String name, String path, String domain) {
      return withCookies(new Http.Cookie(name, "", 0, path, domain, false, true, null, false));
    }

    /**
     * @param request Current request
     * @return The session carried by this WebSocket upgrade response. Reads the given request's
     *     session if this response does not modify the session.
     */
    public Http.Session session(Http.RequestHeader request) {
      if (session != null) {
        return session;
      } else {
        return request.session();
      }
    }

    /**
     * Sets a new session for this WebSocket upgrade response, discarding the existing session.
     *
     * @param session the session to set with this WebSocket upgrade response
     * @return the transformed copy
     */
    public Accepted<In, Out> withSession(Http.Session session) {
      return new Accepted<>(flow, subprotocol, headers, cookies, session);
    }

    /**
     * Sets a new session for this WebSocket upgrade response, discarding the existing session.
     *
     * @param session the session to set with this WebSocket upgrade response
     * @return the transformed copy
     */
    public Accepted<In, Out> withSession(Map<String, String> session) {
      return withSession(new Http.Session(session));
    }

    /**
     * Discards the existing session for this WebSocket upgrade response.
     *
     * @return the transformed copy
     */
    public Accepted<In, Out> withNewSession() {
      return withSession(Collections.emptyMap());
    }

    /**
     * Adds values to this WebSocket upgrade response's session.
     *
     * @param values A map with values to add to this response's session
     * @return the transformed copy
     */
    public Accepted<In, Out> addingToSession(
        Http.RequestHeader request, Map<String, String> values) {
      return withSession(session(request).adding(values));
    }

    /**
     * Adds the given key and value to this WebSocket upgrade response's session.
     *
     * @param key The key to add to this response's session
     * @param value The value to add to this response's session
     * @return the transformed copy
     */
    public Accepted<In, Out> addingToSession(Http.RequestHeader request, String key, String value) {
      Map<String, String> newValues = new HashMap<>(1);
      newValues.put(key, value);
      return addingToSession(request, newValues);
    }

    /**
     * Removes values from this WebSocket upgrade response's session.
     *
     * @param keys Keys to remove from session
     * @return the transformed copy
     */
    public Accepted<In, Out> removingFromSession(Http.RequestHeader request, String... keys) {
      return withSession(session(request).removing(keys));
    }
  }

  /**
   * Utility class for creating WebSockets.
   *
   * @param <In> the type the websocket reads from clients (e.g. String, JsonNode)
   * @param <Out> the type the websocket outputs back to remote clients (e.g. String, JsonNode)
   */
  public static class MappedWebSocketAcceptor<In, Out> {
    private final PartialFunction<Message, F.Either<In, Message>> inMapper;
    private final Function<Out, Message> outMapper;

    public MappedWebSocketAcceptor(
        PartialFunction<Message, F.Either<In, Message>> inMapper,
        Function<Out, Message> outMapper) {
      this.inMapper = inMapper;
      this.outMapper = outMapper;
    }

    /**
     * Accept a WebSocket.
     *
     * @param f A function that takes the request header, and returns a future of either the result
     *     to reject the WebSocket connection with, or a flow to handle the WebSocket messages.
     * @return The WebSocket handler.
     */
    public WebSocket acceptOrResult(
        Function<Http.RequestHeader, CompletionStage<F.Either<Result, Flow<In, Out, ?>>>> f) {
      return WebSocket.acceptOrResult(inMapper, f, outMapper);
    }

    /**
     * Accept a WebSocket with handshake metadata.
     *
     * @param f A function that takes the request header, and returns a future of either the result
     *     to reject the WebSocket connection with, or an accepted WebSocket with its flow and
     *     handshake metadata.
     * @return The WebSocket handler.
     */
    public WebSocket acceptOrResultWithOptions(
        Function<Http.RequestHeader, CompletionStage<F.Either<Result, Accepted<In, Out>>>> f) {
      return WebSocket.acceptOrResultWithOptions(inMapper, f, outMapper);
    }

    /**
     * Accept a WebSocket.
     *
     * @param f A function that takes the request header, and returns a flow to handle the WebSocket
     *     messages.
     * @return The WebSocket handler.
     */
    public WebSocket accept(Function<Http.RequestHeader, Flow<In, Out, ?>> f) {
      return acceptOrResult(
          request -> CompletableFuture.completedFuture(F.Either.Right(f.apply(request))));
    }

    /**
     * Accept a WebSocket with handshake metadata.
     *
     * @param f A function that takes the request header, and returns an accepted WebSocket with its
     *     flow and handshake metadata.
     * @return The WebSocket handler.
     */
    public WebSocket acceptWithOptions(Function<Http.RequestHeader, Accepted<In, Out>> f) {
      return acceptOrResultWithOptions(
          request -> CompletableFuture.completedFuture(F.Either.Right(f.apply(request))));
    }
  }

  /**
   * Helper to create handlers for WebSockets.
   *
   * @param inMapper Function to map input messages. If it produces left, the message will be passed
   *     to the WebSocket flow, if it produces right, the message will be sent back out to the
   *     client - this can be used to send errors directly to the client.
   * @param f The function to handle the WebSocket.
   * @param outMapper Function to map output messages.
   * @return The WebSocket handler.
   */
  private static <In, Out> WebSocket acceptOrResult(
      PartialFunction<Message, F.Either<In, Message>> inMapper,
      Function<Http.RequestHeader, CompletionStage<F.Either<Result, Flow<In, Out, ?>>>> f,
      Function<Out, Message> outMapper) {
    return new WebSocket() {
      @Override
      public CompletionStage<F.Either<Result, Flow<Message, Message, ?>>> apply(
          Http.RequestHeader request) {
        return f.apply(request)
            .thenApply(
                resultOrFlow -> {
                  if (resultOrFlow.left.isPresent()) {
                    return F.Either.Left(resultOrFlow.left.get());
                  } else {
                    Flow<Message, Message, ?> flow =
                        PekkoStreams.bypassWith(
                            Flow.<Message>create().collect(inMapper),
                            play.api.libs.streams.PekkoStreams.onlyFirstCanFinishMerge(2),
                            resultOrFlow.right.get().map(outMapper::apply));
                    return F.Either.Right(flow);
                  }
                });
      }
    };
  }

  private static <In, Out> WebSocket acceptOrResultWithOptions(
      PartialFunction<Message, F.Either<In, Message>> inMapper,
      Function<Http.RequestHeader, CompletionStage<F.Either<Result, Accepted<In, Out>>>> f,
      Function<Out, Message> outMapper) {
    return new WebSocket() {
      @Override
      public CompletionStage<F.Either<Result, Flow<Message, Message, ?>>> apply(
          Http.RequestHeader request) {
        return applyWithOptions(request)
            .thenApply(
                resultOrAccepted -> {
                  if (resultOrAccepted.left.isPresent()) {
                    return F.Either.Left(resultOrAccepted.left.get());
                  } else {
                    return F.Either.Right(resultOrAccepted.right.get().flow());
                  }
                });
      }

      @Override
      public CompletionStage<F.Either<Result, Accepted<Message, Message>>> applyWithOptions(
          Http.RequestHeader request) {
        return f.apply(request)
            .thenApply(
                resultOrAccepted -> {
                  if (resultOrAccepted.left.isPresent()) {
                    return F.Either.Left(resultOrAccepted.left.get());
                  } else {
                    Accepted<In, Out> accepted = resultOrAccepted.right.get();
                    Flow<Message, Message, ?> flow =
                        PekkoStreams.bypassWith(
                            Flow.<Message>create().collect(inMapper),
                            play.api.libs.streams.PekkoStreams.onlyFirstCanFinishMerge(2),
                            accepted.flow().map(outMapper::apply));
                    return F.Either.Right(
                        new Accepted<>(
                            flow,
                            accepted.subprotocol(),
                            accepted.headers(),
                            accepted.cookies(),
                            accepted.session().orElse(null)));
                  }
                });
      }
    };
  }

  private static Optional<String> firstRequestedSubprotocol(Http.RequestHeader request) {
    return request
        .header("Sec-WebSocket-Protocol")
        .flatMap(
            header -> {
              for (String protocol : header.split(",")) {
                String trimmed = protocol.trim();
                if (!trimmed.isEmpty()) {
                  return Optional.of(trimmed);
                }
              }
              return Optional.empty();
            });
  }
}
